[APM] Service Map data background task and API endpoint

ogupte opened the pull request:
Addresses #48996.

Adds data later for the service maps feature in APM:
- creates new index `apm-service-connections`
- schedules periodic task that stores service connection documents in the `apm-service-connections` index
- implements endpoint `/api/apm/service-map` that query/filters service connections to render the service map

<img width="887" alt="Screen Shot 2019-11-26 at 12 13 28 AM" src="https://user-images.githubusercontent.com/1967266/69610917-9f20b300-0fe1-11ea-8e83-ef4cc7050e09.png">

In order to test this out without agents setting `destination.address`, you can install an ingest pipeline that approximates agent support:
```
git clone https://github.com/ogupte/apm-ui-seed.git
cd apm-ui-seed
yarn
bin/apm-ui-service-map-ingest --host='http://localhost:9200' --auth='elastic:password' install
```
when you're done testing, you can uninstall the dev pipeline with:
```
bin/apm-ui-service-map-ingest --host='http://localhost:9200' --auth='elastic:password' uninstall
```

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/10213/)
* Commit: a07a4f8b94af61792e680ce795cfa3ada466c692

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/11518/)
* Commit: 16b45f0dee0cf23d8d0628ea68270bc9b32aae23

ogupte committed:
Adds data later for the service maps feature in APM:
- creates required scripts in elasticsearch
- adds ingest pipeline that sets destination.address from span names
- schedules periodic task that create connection documents
- implements endpoints that query/filters connections to render the service map

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/12090/)
* Commit: 2ea43584dfeaf6e8d265b0a484f34ab6b0c6381c

dgieselaar commented on x-pack/legacy/plugins/apm/server/lib/service_map/initialize_service_maps.ts:
```diff
@@ -0,0 +1,109 @@
+      if (fetchedTasks.docs.length) {
+        await taskManager.remove('servicemap-processor');
+      }
+      await taskManager.schedule({
```
I'm not sure if the point of this is to prevent this task being scheduled more than once, but if that is the case, then you can now use `ensureScheduled`: https://github.com/elastic/kibana/pull/50232

sqren commented on x-pack/legacy/plugins/apm/index.ts:
```diff
@@ -113,6 +126,12 @@ export const apm: LegacyPluginInitializer = kibana => {
+
+      if (config.get('xpack.apm.serviceMapEnabled')) {
+        initializeServiceMaps(server).catch(error => {
+          throw error;
```
This will result in an unhandled promise exception

sqren commented on x-pack/legacy/plugins/apm/server/lib/service_map/get_service_map.ts:
```diff
@@ -0,0 +1,164 @@
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import { idx } from '@kbn/elastic-idx';
```
idx doing a comeback??? :D

sqren commented on x-pack/legacy/plugins/apm/public/components/app/ServiceMap/index.tsx:
```diff
@@ -37,19 +37,51 @@ ${theme.euiColorLightShade}`,
         return callApmApi({
-          pathname: '/api/apm/service-map',
-          params: { query: { start, end } }
+          pathname: '/api/apm/service-map/all',
```
What if a service is called `all`

sqren commented on x-pack/legacy/plugins/apm/server/lib/service_map/get_service_map.ts:
```diff
@@ -0,0 +1,164 @@
+    });
+  }
+
+  // @ts-ignore
```
Is this still needed?

sqren commented on src/legacy/core_plugins/apm_oss/index.js:
```diff
@@ -38,18 +38,24 @@ export default function apmOss(kibana) {
         spanIndices: Joi.string().default('apm-*'),
         metricsIndices: Joi.string().default('apm-*'),
         onboardingIndices: Joi.string().default('apm-*'),
+        apmAgentConfigurationIndex: Joi.string().default('.apm-agent-configuration')
```
Was it intentional to re-add this?

sqren commented on x-pack/legacy/plugins/apm/public/components/app/ServiceMap/index.tsx:
```diff
@@ -37,19 +37,51 @@ ${theme.euiColorLightShade}`,
+              start,
+              end,
+              environment,
+              uiFilters: JSON.stringify(uiFiltersOmitEnv)
```
We normally don't send `environment` like a separate query params but just keep it in `uiFilters`. Can't we do the same here?

sqren commented on x-pack/legacy/plugins/apm/server/routes/services.ts:
```diff
@@ -86,15 +86,48 @@ export const serviceNodeMetadataRoute = createRoute(() => ({
+
 export const serviceMapRoute = createRoute(() => ({
-  path: '/api/apm/service-map',
+  path: '/api/apm/service-map/{serviceName}',
```
As we talked about `/api/apm/service-map/{serviceName}` and `/api/apm/service-map/all` can be combined to a single route if `serviceName` is converted to an optional query param

sqren commented on x-pack/legacy/plugins/apm/server/routes/services.ts:
```diff
@@ -86,15 +86,48 @@ export const serviceNodeMetadataRoute = createRoute(() => ({
-      return getServiceMap();
+  handler: async ({ context, request }) => {
+    if (!context.config['xpack.apm.serviceMapEnabled']) {
+      return new Boom('Not found', { statusCode: 404 });
```
`serviceMapEnabled` should be removed

sqren replied:
... or, are there any reasons to let users disable it? We don't allow this for any other features.

ogupte replied:
yeah we can safely remove this config.

ogupte replied:
after talking with folks, the consensus was to keep the config until FF

sqren replied:
what's the reason to wait until FF if we decided to do this?

sqren commented on x-pack/legacy/plugins/apm/server/lib/service_map/get_service_map.ts:
```diff
@@ -0,0 +1,164 @@
+        };
+      }>;
+    };
+  }> = idx(aggregations, _ => _.conns.buckets) || [];
```
@dgieselaar is this something our magic agg types should be able to infer instead of them being spelled out?

dgieselaar replied:
Looks like this is no longer an issue, correct?

sqren commented on x-pack/legacy/plugins/apm/server/lib/service_map/initialize_service_maps.ts:
```diff
@@ -0,0 +1,109 @@
+      'Unable to set up required scripts for APM Service Maps:'
+    );
+    server.log('error', error);
+  });
```
Is it the intention to continue initialization if `setupRequiredScripts` fails?

ogupte replied:
that was the idea, the result will be some error messages in kibana logs & service maps UI not displaying any nodes

sqren replied:
How's that useful? What about just terminating init as soon as something goes wrong?

sqren commented on x-pack/legacy/plugins/apm/server/lib/service_map/initialize_service_maps.ts:
```diff
@@ -0,0 +1,109 @@
+                term: {
+                  'task.taskType': 'serviceMap'
+                }
+              }
```
Nit: what do you think about single-lining terms like this:

```js
  { term: { _id: "task:servicemap-processor" } },
  { term: { "task.taskType": "serviceMap" } }
```

sqren commented on x-pack/legacy/plugins/apm/server/lib/service_map/run_service_map_task.ts:
```diff
@@ -0,0 +1,248 @@
+
+function interestingTransactions(since?: string, afterKey?: any) {
+  if (!since) {
+    since = 'now-1h';
```
How was this value chosen? Should it be a constant?

sqren replied:
Btw what about adding it to the signature?

sqren commented on x-pack/legacy/plugins/apm/server/lib/service_map/run_service_map_task.ts:
```diff
@@ -0,0 +1,248 @@
+    query: {
+      bool: {
+        filter: [
+          { exists: { field: 'destination.address' } },
```
Use constants

sqren commented on x-pack/legacy/plugins/apm/server/lib/service_map/run_service_map_task.ts:
```diff
@@ -0,0 +1,248 @@
+          { exists: { field: 'destination.address' } },
+          { exists: { field: 'trace.id' } },
+          { exists: { field: 'span.duration.us' } },
+          { range: { '@timestamp': { gt: since } } }
```
What about using `start` and `end` like in the rest of the codebase?

sqren replied:
Am I correct in understanding that `end` is currently ignored?

sqren replied:
Ah, nvm. This is not the function called by the end-user but the job.

sqren commented on x-pack/legacy/plugins/apm/server/lib/service_map/run_service_map_task.ts:
```diff
@@ -0,0 +1,248 @@
+      body
+    });
+  }
+}
```
As we talked about, I think this file would benefit from being refactored into smaller functions, having better types, and being less mutable (if possible).

sqren commented on x-pack/legacy/plugins/apm/server/lib/service_map/setup_required_scripts.ts:
```diff
@@ -0,0 +1,298 @@
+
+            return new ArrayList(conns)
+          }
+          return [];
```
Do we have tests for this? I think that's quiet important since we have no static safety - not even syntax highlighting - to notify us of any issues.

sqren replied:
Btw ran it through a formatter to fix some minor indentation and spacing issues:`
```java
void extractChildren(def caller, def spans, def upstream, def conns, def count) {
  // TODO: simplify this
  if (spans.containsKey(caller.id)) {
    for (s in spans[caller.id]) {
      if (caller.span_type == 'external') {
        upstream.add(caller.service_name + "/" + caller.environment);
        def conn = new HashMap();
        conn.caller = caller;
        conn.callee = s;
        conn.upstream = new ArrayList(upstream);
        conns.add(conn);
        extractChildren(s, spans, upstream, conns, count);
        upstream.remove(upstream.size() - 1);
      } else {
        extractChildren(s, spans, upstream, conns, count);
      }
    }
  } else {
    // no connection found
    def conn = new HashMap();
    conn.caller = caller;
    conn.upstream = new ArrayList(upstream);
    conn.upstream.add(caller.service_name + "/" + caller.environment);
    conns.add(conn);
  }
}

def conns = new HashSet();
def spans = new HashMap();

// merge results from shards
for (state in states) {
  for (s in state.entrySet()) {
    def v = s.getValue();
    def k = s.getKey();
    if (!spans.containsKey(k)) {
      spans[k] = v;
    } else {
      for (p in v) {
        spans[k].add(p);
      }
    }
  }
}

if (spans.containsKey(null) && spans[null].size() > 0) {
  def node = spans[null][0];
  def upstream = new ArrayList();
  extractChildren(node, spans, upstream, conns, 0);
  return new ArrayList(conns)
}

return [];
```

sqren commented on x-pack/legacy/plugins/apm/server/lib/service_map/setup_required_scripts.ts:
```diff
@@ -0,0 +1,298 @@
+        source: `
+          void extractChildren(def caller, def spans, def upstream, def conns, def count) {
+            // TODO: simplify this
+            if (spans.containsKey(caller.id)) {
```
Is it still the plan to simplify this?

sqren commented on x-pack/legacy/plugins/apm/server/lib/service_map/setup_required_scripts.ts:
```diff
@@ -0,0 +1,298 @@
+            // TODO: simplify this
+            if (spans.containsKey(caller.id)) {
+              for(s in spans[caller.id]) {
+                  if (caller.span_type=='external') {
```
I see `external` being used a couple of places. Should it be a constant like we have for [transaction types](https://github.com/elastic/kibana/blob/45ef370e8443baf7eeec5bc2cb7e25cae31fff10/x-pack/legacy/plugins/apm/common/transaction_types.ts#L9)?

sqren commented on x-pack/legacy/plugins/apm/server/lib/service_map/setup_required_scripts.ts:
```diff
@@ -0,0 +1,298 @@
+            if: "ctx.span != null && ctx.span.type == 'ext'",
+            field: 'span.type',
+            value: 'external'
+          }
```
Looks like we rewrite `ext` to `external`. Why is that?

ogupte replied:
At the time, the value was not consistent, so this normalized everything to `external`.

see https://github.com/elastic/apm-agent-nodejs/issues/1225#issue-469295459

ogupte replied:
should also note that this ingest pipeline will not be merged.

sqren commented on x-pack/legacy/plugins/apm/server/lib/service_map/setup_required_scripts.ts:
```diff
@@ -0,0 +1,298 @@
+    // get current apm ingest pipeline
+    apmIngestPipeline = await getIngestPipelineApm(callCluster);
+  } catch (error) {
+    if (error.statusCode !== 404) {
```
Maybe add a comment about why we are swallowing 404 errors

sqren replied:
nvm. comment below is sufficient 👍

sqren commented on x-pack/plugins/apm/server/index.ts:
```diff
@@ -33,6 +36,9 @@ export function mergeConfigs(apmOssConfig: APMOSSConfig, apmConfig: APMXPackConf
     'xpack.apm.ui.maxTraceItems': apmConfig['ui.maxTraceItems'],
     'xpack.apm.ui.transactionGroupBucketSize': apmConfig['ui.transactionGroupBucketSize'],
     'xpack.apm.autocreateApmIndexPattern': apmConfig.autocreateApmIndexPattern,
+    'xpack.apm.serviceMapIndexPattern': apmConfig.serviceMapIndexPattern,
```
`serviceMapIndexPattern` is misleading. It's not an index pattern (in the Kibana sense). This should follow the names of the other keys: `serviceMapIndices`.
Even though this is an xpack feature I'm contemplating whether it'll make more sense to keep this in `apm_oss` where all other indices are configured.

ogupte replied:
I don't think it belongs in the apm_oss set of configs, since it's a premium feature. it will not be accessed if the use doesn't pass a license check.

sqren replied:
Yeah, I know, but it feels weird as an xpack user that you have to configure your indices differently. Let's get a second opinion on this.

sqren replied:
@bmorelli25 i'm interested in your take on this. Currently all of the apm indicies are namespaced under `apm.oss`. Examples: `apm_oss.errorIndices`, `apm_oss.transactionIndices` and `apm_oss.spanIndices`. Now we are also going to add a setting for Service Maps indices. 

From a docs perspective what do you think makes more sense to: 
 - `apm_oss.serviceMapIndices` (consistency with other index configuration option names)
 - `xpack.apm.serviceMapIndices` (Service Maps are only available in basic so is not used for anything in oss)

ogupte replied:
>(Service Maps are only available in basic...)
should only be available in platinum

bmorelli25 replied:
My vote is for `xpack.apm.serviceMapIndices`.

> it feels weird as an xpack user that you have to configure your indices differently.

I agree, it does feel a little weird, but I think it's less weird than namespacing a platinum feature under the `oss` prefix.

sqren replied:
Makes sense. Then let's stick to `xpack.apm.serviceMapIndices`.

sqren commented on x-pack/plugins/apm/server/index.ts:
```diff
@@ -33,6 +36,9 @@ export function mergeConfigs(apmOssConfig: APMOSSConfig, apmConfig: APMXPackConf
     'xpack.apm.autocreateApmIndexPattern': apmConfig.autocreateApmIndexPattern,
+    'xpack.apm.serviceMapIndexPattern': apmConfig.serviceMapIndexPattern,
+    'xpack.apm.serviceMapDestinationIndex': apmConfig.serviceMapDestinationIndex,
+    'xpack.apm.serviceMapDestinationPipeline': apmConfig.serviceMapDestinationPipeline,
```
What's the purpose of making this configurable?

ogupte replied:
it is to optionally provide a pipeline for the bulk-created docs with connection data. I'm doubtful this is really necessary, so it can safely be removed.

sqren commented on x-pack/plugins/apm/server/index.ts:
```diff
@@ -33,6 +36,9 @@ export function mergeConfigs(apmOssConfig: APMOSSConfig, apmConfig: APMXPackConf
     'xpack.apm.ui.transactionGroupBucketSize': apmConfig['ui.transactionGroupBucketSize'],
     'xpack.apm.autocreateApmIndexPattern': apmConfig.autocreateApmIndexPattern,
+    'xpack.apm.serviceMapIndexPattern': apmConfig.serviceMapIndexPattern,
+    'xpack.apm.serviceMapDestinationIndex': apmConfig.serviceMapDestinationIndex,
```
Who will make queries against this index? Internal kibana user, end user or a third user? If it's the internal kibana user we'll run into the same problems we did with ACM where the built-in role cannot be configured. So if someone changes it from the default service maps will stop working.

ogupte replied:
it is used to optionally place bulk-created documents which have the connections processed by the task runner, so it would be the kibana user performing this operation on the index.

sqren replied:
Do you agree that this would break if the end-user uses a custom index?

ogupte replied:
as long as they customize`serviceMapDestinationIndex` and `serviceMapIndexPattern`, to match, then it won't break the feature. We query `serviceMapIndexPattern` when we load the service maps UI, so as long as there is connection data there, it will return results.

sqren replied:
Afaict if the user changes `serviceMapDestinationIndex` the kibana internal user will not be able to access it. And the feature will break. This was the problems we were having with ACM.
How is this different?

ogupte committed:
- fix UI bug where service map data did not update correctly
- replace this.kbnServer.afterPluginsInit and task object fetch with taskManager.ensureScheduled
- simplify routes: move serviceName to query param and consolidate endpoints
- clean up code in get_service_map.ts, constants, typescript fixes
- add typescript support for composite aggregations

sqren commented on x-pack/legacy/plugins/apm/server/lib/service_map/run_service_map_task.ts:
```diff
@@ -0,0 +1,250 @@
+  while (true) {
+    const q = interestingTransactions(lastRun, afterKey);
+    const txs = await callCluster('search', {
+      index: apmIdxPattern,
```
Shouldn't this query only search transaction indices? Currently it searches all indices (errors, spans, metrics, transactions)

ogupte replied:
right now, it it includes connections derived from trace ids of both transactions and spans within a specific time range. we could limit this to transactions only and it may show fewer connections in the results, but eventually as more time passes and a missing trace's transaction comes within the latest time interval, those connections will be included in the service map data. I'll do what you suggest and see what happens.

sqren replied:
If you want to search both transactions and spans, I'd think you can specify that like:
`index: [transactionIndices, spanIndices]`

sqren commented on x-pack/legacy/plugins/apm/server/lib/service_map/get_service_map.ts:
```diff
@@ -0,0 +1,173 @@
+        source: string;
+        target: string;
+      };
+}
```
What do you think about making it explicit that this is cytoscape specific?
```ts
interface CyNode {
  id: string;
}

interface CyEdge {
  id: string;
  source: string;
  target: string;
}

interface CyElement {
  data: CyNode | CyEdge;
}
```

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13384/)
* Commit: 7ec60bec2fa3b0682d574dbbf8726b70ab3c2150

sqren commented on x-pack/legacy/plugins/apm/server/lib/service_map/run_service_map_task.ts:
```diff
@@ -0,0 +1,250 @@
+    const traces = new Set<string>();
+
+    txs.aggregations['ext-conns'].buckets.forEach((bucket: any) => {
+      bucket.smpl.tracesample.hits.hits.forEach((hit: any) => {
```
We should have proper types here

dgieselaar replied:
I'm adding support for `composite` aggregations in https://github.com/elastic/kibana/pull/51612. Should be possible for @ogupte to copy and paste the relevant bits, LMK if that doesn't work out.

ogupte replied:
i already added support for `composite`, `diversified_sampler`, and `scripted_metric` aggregations in this PR

ogupte replied:
I'll definitely replace my composite stuff with yours, your key inference it very nice :)

sqren commented on x-pack/legacy/plugins/apm/server/lib/service_map/get_service_map.ts:
```diff
@@ -0,0 +1,173 @@
+      ? environment === ENVIRONMENT_NOT_DEFINED
+        ? 'null'
+        : environment
+      : '*';
```
As we talked about perhaps it's better to remove the option to query for services without environment.

ogupte replied:
The problem is that this is the global environment filter, which includes `Not defined` as an option. To remove it would affect other screens as well.

ogupte replied:
but instead of using the string `'null'` to mark an environment which is not defined, i changed it to `ENVIRONMENT_NOT_DEFINED`, so that this nested ternary is no longer necessary.

sqren replied:
> To remove it would affect other screens as well.

yeah, I thought that's what we agreed - removing it everywhere? We should get 👍 from @formgeist but I don't see that being a problem.

formgeist replied:
@sqren @ogupte I was under the impression early on that we'd have to remove the query bar, but I think that was even before we added the environment filter in there. Can you fill me in, what's the reason that we cannot have the "not defined" option in the environment filter in this case?

ogupte committed:
- rewrite the task run logic to be more readable and less procedural
- broke up the task into more files for better organization and testing

ogupte committed:
fix bug in query and timestamp

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13901/)
* Commit: 7a812ae1379ec34a4d51c77efab953a5e6ca748c

sorenlouv reviewed (commented):
@ogupte Good progress 👍  Added a bunch of questions. 

Additional question: where is the mapping for the service map index located?

roncohen commented on x-pack/legacy/plugins/apm/server/lib/service_map/map_service_connection_to_bulk_index.ts:
```diff
@@ -0,0 +1,56 @@
+      connection: {
+        upstream: {
+          list: serviceConnection.upstream,
+          keyword: serviceConnection.upstream.join('->') // TODO is this even used/necessary?
```
I used this at some point when we were OK to show an upstream service multiple times if it has multiple paths into the current service. I _think_ i ended up not using it. Maybe @smith can comment?

smith replied:
I don't think it is used.

ogupte committed:
Include scripted_metric scripts with the query rather than persisting them in ES.

ogupte committed:
replace field strings with shared constants

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14417/)
* Commit: ec845675f89142b7e7b4a1fb5b1e90a27b139916

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14417/","number":"14417","commit":"ec845675f89142b7e7b4a1fb5b1e90a27b139916"}]}
PIPELINE-->

ogupte committed:
Addressed feedback and fixed timestamp snapshot

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14454/)
* Commit: 06b07ee2a7baf5d62fb7813a154a45adbb7f3793

### History
* :broken_heart: [Build #14417](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14417/) failed ec845675f89142b7e7b4a1fb5b1e90a27b139916

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14417/","number":"14417","commit":"ec845675f89142b7e7b4a1fb5b1e90a27b139916"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14454/","number":"14454","commit":"06b07ee2a7baf5d62fb7813a154a45adbb7f3793"}]}
PIPELINE-->

sqren commented on x-pack/legacy/plugins/apm/common/elasticsearch_fieldnames.test.ts:
```diff
@@ -13,7 +13,7 @@ import * as fieldnames from './elasticsearch_fieldnames';
 describe('Transaction', () => {
   const transaction: AllowUnknownProperties<Transaction> = {
-    '@timestamp': new Date().toString(),
+    '@timestamp': new Date(1337).toString(),
```
nit: I don't think it matters much, but `@timestamp` is ISO 8601 format. We can get the same with `new Date(1337).toUTCString()`

ogupte committed:
creates apm-service-connections index and adds it to getApmIndices

sqren commented on x-pack/legacy/plugins/apm/server/routes/create_apm_api.ts:
```diff
@@ -75,6 +75,7 @@ const createApmApi = () => {
     .add(serviceTransactionTypesRoute)
     .add(servicesRoute)
     .add(serviceNodeMetadataRoute)
+    .add(serviceMapRoute)
```
I think this should have its own section

sqren commented:
> @sqren @ogupte I was under the impression early on that we'd have to remove the query bar, but I think that was even before we added the environment filter in there. Can you fill me in, what's the reason that we cannot have the "not defined" option in the environment filter in this case?

The "not defined" option is adding some complexity around the code base that I'd rather be without. It's quite easy to mix up the logic for "All environments" and "Not defined". We could embrace the complexity by adding additional test coverage but I'm doubtful of the value of "not defined" in the first place.

If someone doesn't use environments they can list all their services with "All environments". 
The only use case I can think of, is if someone uses environments for all but a few services and want to list the services without environments. This can be achieved with the query bar: `not service.environment : *`.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14475/)
* Commit: a5edd893e54fc7de40768b42e604e850a602fb77

### History
* :broken_heart: [Build #14454](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14454/) failed 06b07ee2a7baf5d62fb7813a154a45adbb7f3793
* :broken_heart: [Build #14417](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14417/) failed ec845675f89142b7e7b4a1fb5b1e90a27b139916

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14417/","number":"14417","commit":"ec845675f89142b7e7b4a1fb5b1e90a27b139916"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14454/","number":"14454","commit":"06b07ee2a7baf5d62fb7813a154a45adbb7f3793"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14475/","number":"14475","commit":"a5edd893e54fc7de40768b42e604e850a602fb77"}]}
PIPELINE-->

sqren commented on x-pack/legacy/plugins/apm/common/service_map_constants.ts:
```diff
@@ -0,0 +1,8 @@
+ */
+
+export const SERVICE_MAP_TASK_TYPE = 'apmServiceMap';
+export const SERVICE_MAP_TASK_ID = 'apm-servicemap-processor';
```
Should we have a convention around camelCase vs kebab-case for constants?

ogupte replied:
We already use snake case for constants (eg `SERVICE_MAP_TASK_ID`) as for the value, it's  depends on the context. There's no existing convention for task types and task ids. Some plugins mix snake and kebab case for id: `plugin_task-id`. from what i see, it looks like other plugins tend to use snake case for the task type. We could change the type to `apm_service_map` to conform if we want to. The task type is represented in the API by an object key in the task definition, so i don't recommend kebab case.

sqren commented on x-pack/legacy/plugins/apm/index.ts:
```diff
@@ -11,10 +11,17 @@ import { APMPluginContract } from '../../../plugins/apm/server/plugin';
+      'elasticsearch',
+      'xpack_main',
+      'apm_oss',
+      'task_manager'
```
This is another reason for not having `serviceMapEnabled` anymore: APM requires `task_manager` regardless.

sqren commented on x-pack/legacy/plugins/apm/public/components/app/ServiceMap/index.tsx:
```diff
@@ -37,19 +37,36 @@ ${theme.euiColorLightShade}`,
+            query: {
+              start,
+              end,
+              environment,
```
I think my previous comment was lost in a rebase: 
Is it not possible to extract `environment` from `uiFilters` on the backend?

sqren commented on x-pack/legacy/plugins/apm/server/lib/helpers/es_client.ts:
```diff
@@ -165,3 +166,18 @@ export function getESClient(
 }
+
+export type SearchClient = ReturnType<typeof getSearchClient>;
+export function getSearchClient(callCluster: CallCluster) {
```
Shouldn't this be called `getInternalEsClient` to be consistent with `getESClient` and `getInternalSavedObjectsClient`?

```suggestion
export function getInternalEsClient(callCluster: CallCluster) {
```

sqren replied:
Actually, `getESClient` only relies on `request` or debugging purposes. Could you use that instead of making a new client handler?

sqren replied:
After playing around with this for a bit I came up with this:
```ts
function getESClient(
  savedObjectsClient: SavedObjectsClientContract,
  uiSetttingsClient: UiSettingsClient,
  esClient:
    | IScopedClusterClient['callAsCurrentUser']
    | IScopedClusterClient['callAsInternalUser'],
  config: APMConfig,
  onRequest: (endpoint: string, params: Record<any, any>) => void
) 
```

All clients are explicitly passed in. This way makes its up to the consumer to decide whether the internal user or the request user should be used. I've replaced the `request` argument with an `onRequest` callback that can be used to log elasticsearch queries.

sqren replied:
This will also affect any downstream dependency that depends on `context` like `getParamsForSearchRequest`:

**Before**
```ts
async function getParamsForSearchRequest(
  context: APMRequestHandlerContext,
  params: ESSearchRequest,
  apmOptions?: APMOptions
) {
```

**After**
```ts
async function getParamsForSearchRequest(
  savedObjectsClient: SavedObjectsClientContract,
  uiSettingsClient: IUiSettingsClient,
  config: APMConfig,
  params: ESSearchRequest,
  apmOptions?: APMOptions
) {
```

@dgieselaar any opinions on something like this?

sqren replied:
Full example of `getEsClient`
```ts
export function getESClient(
  apmIndices: ApmIndicesConfig,
  uiSettingsClient: IUiSettingsClient,
  esClient:
    | IScopedClusterClient['callAsCurrentUser']
    | IScopedClusterClient['callAsInternalUser'],
  onRequest: (endpoint: string, params: Record<any, any>) => void = () => {}
) {
  return {
    search: async <
      TDocument = unknown,
      TSearchRequest extends ESSearchRequest = {}
    >(
      params: TSearchRequest,
      apmOptions?: APMOptions
    ): Promise<ESSearchResponse<TDocument, TSearchRequest>> => {
      const nextParams = await getParamsForSearchRequest(
        apmIndices,
        uiSettingsClient,
        params,
        apmOptions
      );

      onRequest('search', nextParams);

      return (esClient('search', nextParams) as unknown) as Promise<
        ESSearchResponse<TDocument, TSearchRequest>
      >;
    },
    index: <Body>(params: APMIndexDocumentParams<Body>) => {
      onRequest('index', params);
      return esClient('index', params);
    },
    delete: (params: IndicesDeleteParams) => {
      onRequest('delete', params);
      return esClient('delete', params);
    },
    indicesCreate: (params: IndicesCreateParams) => {
      onRequest('indices.create', params);
      return esClient('indices.create', params);
    }
  };
}
```

ogupte replied:
OK i made this change in the latest version, the only other change was in setup_request, so it wasn't too invasive

ogupte committed:
generalize getEsClient to leverage it in run_service_map_task

ogupte committed:
remove development helper ingest pipeline which extracted destination.address from span names

ogupte committed:
fix tests to snapshot ISO 8601 timestamps

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14704/)
* Commit: 590ceba4a828e42bf68fce7ff53062fba5bf8dca

### History
* :broken_heart: [Build #14475](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14475/) failed a5edd893e54fc7de40768b42e604e850a602fb77
* :broken_heart: [Build #14454](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14454/) failed 06b07ee2a7baf5d62fb7813a154a45adbb7f3793
* :broken_heart: [Build #14417](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14417/) failed ec845675f89142b7e7b4a1fb5b1e90a27b139916

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14417/","number":"14417","commit":"ec845675f89142b7e7b4a1fb5b1e90a27b139916"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14454/","number":"14454","commit":"06b07ee2a7baf5d62fb7813a154a45adbb7f3793"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14475/","number":"14475","commit":"a5edd893e54fc7de40768b42e604e850a602fb77"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14704/","number":"14704","commit":"590ceba4a828e42bf68fce7ff53062fba5bf8dca"}]}
PIPELINE-->

ogupte committed:
- code improvements to es scripts
- conform to ESC fields
- performance improvments to get service connections
- add checks for empty bulk requests

ogupte committed:
fix import path failures

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/15584/)
* Commit: 02870016934cfb3ef73f2fcbb63a63c20af5005f

### History
* :broken_heart: [Build #14704](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14704/) failed 590ceba4a828e42bf68fce7ff53062fba5bf8dca
* :broken_heart: [Build #14475](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14475/) failed a5edd893e54fc7de40768b42e604e850a602fb77
* :broken_heart: [Build #14454](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14454/) failed 06b07ee2a7baf5d62fb7813a154a45adbb7f3793
* :broken_heart: [Build #14417](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14417/) failed ec845675f89142b7e7b4a1fb5b1e90a27b139916

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14417/","number":"14417","commit":"ec845675f89142b7e7b4a1fb5b1e90a27b139916"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14454/","number":"14454","commit":"06b07ee2a7baf5d62fb7813a154a45adbb7f3793"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14475/","number":"14475","commit":"a5edd893e54fc7de40768b42e604e850a602fb77"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14704/","number":"14704","commit":"590ceba4a828e42bf68fce7ff53062fba5bf8dca"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/15584/","number":"15584","commit":"02870016934cfb3ef73f2fcbb63a63c20af5005f"}]}
PIPELINE-->

ogupte committed:
fixes typescript and task api errors

sqren commented on x-pack/legacy/plugins/apm/server/lib/service_map/get_service_map.ts:
```diff
@@ -0,0 +1,165 @@
+
+    params.body.query.bool.filter.push({
+      wildcard: {
+        [CONNECTION_UPSTREAM_LIST]: `${upstreamServiceName}/${upstreamEnvironment}`
```
Maybe this discussion has come up before but how much time have we spent looking into having `service.name` and `service.environment` in separate fields?
If the problem is due to querying (because it's an array of objects) could we used nested fields instead?

That would make querying faster and imo simpler (although nested fields does introduce a slightly different syntax)

sqren replied:
@dgieselaar any opinion on this?

dgieselaar replied:
perhaps do both? `service.name`, `service.environment`, and `service.identifier` (or something more appropriately named). I'm not sure if we should use nested, aggregations become more complicated and it doesn't work with instance sorting, which we will probably not use in practice but I think the use case for nested needs to be really compelling as there are some drawbacks.

sqren replied:
My main concern with the current approach is that it is not longer ECS compliant so we can't apply ui filters like normally. Additionally `wildcard` can be terribly slow.

sqren replied:
I'm fine with having both for possible future use cases. Although currently I don't think we want to query on `CONNECTION_UPSTREAM_LIST`

dgieselaar replied:
I'll have a proper look now, might be missing some context

sqren commented on x-pack/legacy/plugins/apm/server/lib/service_map/run_service_map_task.ts:
```diff
@@ -0,0 +1,126 @@
+      )
+    );
+    return traceConnections.flatMap(
+      mapTraceToBulkServiceConnection(apmIndices, servicesInTrace)
```
nit: Making `mapTraceToBulkServiceConnection` a higher-order function seems slightly unnecessary. Perhaps just:
```ts
return traceConnections.flatMap(traceConnection => mapTraceToBulkServiceConnection(apmIndices, servicesInTrace, traceConnection)
```
```

sqren commented on x-pack/legacy/plugins/apm/server/lib/service_map/get_next_transaction_samples.ts:
```diff
@@ -0,0 +1,117 @@
+  esClient: ESClient;
+}) {
+  const params = {
+    index: targetApmIndices,
```
It's not immediately possible to see which indices are being queried. Everywhere else `apmIndices` is passed in. Can we do the same here?
```suggestion
    index: [apmIndices['apm_oss.transactionIndices'], apmIndices['apm_oss.spanIndices']]
```

sqren commented on x-pack/legacy/plugins/apm/server/lib/service_map/get_service_connections.ts:
```diff
@@ -0,0 +1,74 @@
+  esClient: ESClient;
+}) {
+  const params = {
+    index: targetApmIndices,
```
Same as elsewhere

sqren commented on x-pack/legacy/plugins/apm/server/lib/service_map/get_next_transaction_samples.ts:
```diff
@@ -0,0 +1,117 @@
+                }
+              },
+              aggs: {
+                tracesample: {
```
```suggestion
                trace_sample: {
```

sqren commented on x-pack/legacy/plugins/apm/server/lib/service_map/service-connection-es-scripts.ts:
```diff
@@ -0,0 +1,120 @@
+      'span.subtype',
+      'service.name',
+      'service.environment',
+      'destination.address'
```
What do you think about using constants here?

ogupte replied:
I considered it but felt it was best to avoid interpolation of the script source. I didn't want it to look like i was generating dynamic scripts from our code.

sqren commented on x-pack/legacy/plugins/apm/server/lib/service_map/service-connection-es-scripts.ts:
```diff
@@ -0,0 +1,120 @@
+  source: `return state.mappedDocs`
+};
+
+export const reduceServiceConnsScript = {
```
Can you add a comment for each of these scripts about what they are doing?

ogupte committed:
Modify flow of the service map initialization to handle less permissive
default security settings. The kibana user is responsible for creating
and index to the `apm-service-connections` data index, while the apm
user is resposible for kicking off the scheduled task and reading from
`apm-*` indices.

kibanamachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/17069/)
* Commit: f0aec813a2a8c004abe09ad56f0be00b860a3c30

### History
* :broken_heart: [Build #15584](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/15584/) failed 02870016934cfb3ef73f2fcbb63a63c20af5005f
* :broken_heart: [Build #14704](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14704/) failed 590ceba4a828e42bf68fce7ff53062fba5bf8dca
* :broken_heart: [Build #14475](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14475/) failed a5edd893e54fc7de40768b42e604e850a602fb77
* :broken_heart: [Build #14454](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14454/) failed 06b07ee2a7baf5d62fb7813a154a45adbb7f3793
* :broken_heart: [Build #14417](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14417/) failed ec845675f89142b7e7b4a1fb5b1e90a27b139916

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14417/","number":"14417","commit":"ec845675f89142b7e7b4a1fb5b1e90a27b139916"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14454/","number":"14454","commit":"06b07ee2a7baf5d62fb7813a154a45adbb7f3793"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14475/","number":"14475","commit":"a5edd893e54fc7de40768b42e604e850a602fb77"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/14704/","number":"14704","commit":"590ceba4a828e42bf68fce7ff53062fba5bf8dca"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/15584/","number":"15584","commit":"02870016934cfb3ef73f2fcbb63a63c20af5005f"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/17069/","number":"17069","commit":"f0aec813a2a8c004abe09ad56f0be00b860a3c30"}]}
PIPELINE-->

ogupte commented:
closing this in favor of the runtime implementation https://github.com/elastic/kibana/pull/54027
